--- a/root/do_update.sh	1969-12-31 16:05:47.000000000 -0800
+++ b/root/do_update.sh	2017-10-20 02:52:21.000000000 -0700
@@ -1,332 +1,673 @@
 #!/bin/sh
 
-# Input parameters;
-# $1: upgrade or recovery
-# $2: tarball source partition or folder
-# $3: destination partition
-#    - /dev/nandd: main rootfs partition for both upgrade and recovery
-
-###############################################################################
-# How software upgrade works and it consists of below procedures:
-# 1) Mount partitions
-#    - $1, tarball source partition
-#    - /dev/nandh, cache for intermediate data during upgrade process
-# 2) Check if upgrade package (ch_fw__0.1_309.tar.gz) exists
-#    at /root/udisk for upgrade
-#    at /root/
-# 3) Unzip and check integrity of upgrade package by md5sum
-# 4) Update rootfs
-# 5) Update kernel
-# 6) Copy user data
-# 7) NOTE: for Cherry, we can not update u-boot it is inaccessible for linux.
-#    Currently u-boot can only be updated via PhoenixSuit.
-# 8) Remove the intermediate files in cache partition
-# 9) Remove upgrade package from /mnt/update (/dev/mmcblk0p4)
-# 10) Umount the partitions
-###############################################################################
+############################################################################################
+#
+# BashBunny Firmware Recovery and Updater Script - v2.0 - Hak5 2017
+# 
+# Input Arguments:
+#   $1 : upgrade or recovery
+#   $2 : tarball source
+#   $3 : destination
+#
+# Notes:
+#   - /dev/nandf: udisk
+#   - /dev/nandd: rootfs
+#   - /dev/nandh: cache
+#   - u-boot not upgradable via Linux but via PhoenixSuit only.
+#
+#   Upgrade stages:
+#   1) Mount $2 (Source Partition) and /dev/nandh (Cache Partition) 
+#   2) Check if upgrade file exists (ch_fw_*.tar.gz) at udisk.
+#   3) Unpack upgrade file and verify md5sum.
+#   4) Update the rootfs
+#   5) Update the kernel
+#   6) Copy user data
+#   7) Remove intermerdiate files in /dev/nandh
+#   8) Remove the upgrade package from /mnt/update
+#   9) Unmount parititions and reboot.
+############################################################################################
 
 purpose=$1
-sourcepart=$2
-destpart=$3
-cachepart=/dev/nandh
-bootcountpart=/dev/nanda
-kernel_part=/dev/nandc
+sourcepartition=$2
+destpartition=$3
+bootcountpartition=/dev/nanda
+cachepartition=/dev/nandh
+kernelpartition=/dev/nandc
 
 sourcefolder=/root/udisk
-mkdir -p $sourcefolder
 destfolder=/root/rootfs
-mkdir -p $destfolder
 cachefolder=/root/cache
-mkdir -p $cachefolder
 bootcountfolder=/root/bootcount
+mkdir -p $sourcefolder
+mkdir -p $destfolder
+mkdir -p $cachefolder
 mkdir -p $bootcountfolder
 
-echo "========================================================"
-echo "--- We are " $purpose, source partition = $sourcepart, dest partition = $destpart
+touch /root/update.log
+echo " " >> /root/update.log
 
-do_mount_upgrade() {
-	echo
-	sync
-	umount $bootcountpart
-	umount $cachepart
-	umount $destfolder
-
-	echo "--- 1) Mount partitions"
-	case $purpose in
-	upgrade)
-		echo "--- We are upgrading ..."
-		umount $sourcepart
-		# we do not know the file system of mass storage partition
-		mount $sourcepart $sourcefolder
-
-		# upgrade process always formats and clears bootcnt.txt
-		/sbin/mkfs.fat -F 16 -n BOOTCNT $bootcountpart
-		mount -t vfat $bootcountpart $bootcountfolder
-		echo 0 > $bootcountfolder/bootcnt.txt
-		;;
-	recovery)
-		echo "--- We are recovering ..."
-		sourcefolder=/root
-
-		# recovery process always formats and clears bootcnt.txt
-		/sbin/mkfs.fat -F 16 -n BOOTCNT $bootcountpart
-		mount -t vfat $bootcountpart $bootcountfolder
-		echo 0 > $bootcountfolder/bootcnt.txt
-		;;
-	*)
-		echo "--- ERROR: must be either upgrade or recovery !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! "
-		exit 1
-		;;
-	esac
+echo "==============Starting Update==============" >> /root/update.log
+echo "--- We are doing ${purpose}, Source is ${sourcepartition}, Destination is ${destpartition}"
+echo
+
+do_clear_bootcounter() {
+    echo "--- Clearing boot counter ..."
+    /sbin/mkfs.fat -F 16 -n BOOTCNT $bootcountpartition
+    mount -t vfat $bootcountpartition $bootcountfolder
+    echo 0 > $bootcountfolder/bootcnt.txt
 }
 
+do_mount_upgrade() {
+    sync
+    umount $bootcountfolder
+    umount $cachefolder
+    umount $destfolder
+
+    echo "--- Stage 1: Mounting Partitions"
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 1" >> /root/update.log
+    case $purpose in
+        upgrade)
+            echo "--- We are upgrading firmware ..."
+            umount $sourcepartition
+            mount $sourcepartition $sourcefolder
+
+            # Clear the boot counter.
+            do_clear_bootcounter
+            ;;
+        recovery)
+            echo "--- We are recovering firmware ..."
+            sourcefolder=/root
+
+            # Clear the boot counter.
+            do_clear_bootcounter
+            ;;
+        *)
+            echo "--- Fatal Error: input is not upgrade or recovery! Exiting..."
+            exit 1
+            ;;
+    esac
+}
 
 do_unmount() {
-	echo "--- Flushing everything ..."
-	cd
-	sync
-	umount $bootcountpart
-	umount $cachepart
-	umount $destfolder
-
-	case $purpose in
-	upgrade)
-		umount $sourcepart
-		;;
-	recovery)
-		;;
-	*)
-		exit 1
-		;;
-	esac
+    echo "--- Unmounting everything ..."
+    cd
+    sync
+    umount $bootcountpartition
+    umount $cachepartition
+    umount $destfolder
+
+    case $purpose in
+        upgrade)
+            umount $sourcepartition
+            ;;
+    esac
 }
 
 do_exit() {
-	do_unmount
-	exit
+    echo "--- Exiting script and writing log file ..."
+    mv /root/update.log $sourcefolder
+    do_unmount
+    exit
 }
 
 
-###############################################################################
-# 2) Check if upgrade package (ch_fw_1.0_309.tar.gz) exits
-#    at $sourcefolder (originally /dev/nandf) and version number in its filename
-###############################################################################
+########################################################################
+# Stage 2: Check if upgrade package exists in $sourcefolder.
+########################################################################
 
 check_updatefile() {
-	updateFileFound=0
-	echo
-	echo "--- 2) Checking if upgrade package file exists or not ..."
-
-	# convert to upper-case
-#	Below command tr does not work with Debian initrd.
-#	strCherry=`echo "ch_fw_" | tr '[:lower:]' '[:upper:]'`
-	strCherry="ch_fw_"
-	lenCherry=`expr length "$strCherry"`
-#	strCurrentVersion=`cat /etc/cherry-version | tr '[:lower:]' '[:upper:]'`
-
-	# check if file ch_fw_x.x_xxx.tar.gz exits
-	for entry in `ls ${sourcefolder}`; do
-#		entryUpper=`echo $entry | tr '[:lower:]' '[:upper:]'`
-		entryUpper=$entry
-		lenMatch=`expr match "$entryUpper" "$strCherry"`
-		echo "--- entryUpper =" $entryUpper, strCherry = $strCherry, lenCherry = $lenCherry, lenMatch = $lenMatch
-
-		if [ $lenMatch -eq $lenCherry ]; then
-			updateFileFound=1
-# Allow downgrade Software
-#			if [ "$entryUpper" \> "$strCurrentVersion" ]; then
-				updateFileFound=1
-				# break if first matched filename found
-				break
-#			fi
-		fi
-	done
-
-	if [ $updateFileFound -eq 0 ]; then
-		echo "--- Upgrade file is not found. Exit."
-		echo 1 > $bootcountfolder/tar_fail.txt
-		do_exit
-	else
-		echo "--- Upgrade file found:" $entry
-	fi
-}
-
-
-###############################################################################
-# 3) Unzip and check integrity of upgrade package by md5sum
-###############################################################################
-
-folderUpdate=$cachefolder/update
-kernel_filename="uImage"
-rootfs_filename="cherry.rootfs.tar"
-initrd_filename="cherry.initrd.tar"
-ifc_version_filename="cherry-version"
-
-#	Input: $1 = filename
-check_md5() {
-	str_md5_calc=`md5sum $folderUpdate/$1 | cut -d' ' -f1`
-	if [ -f $folderUpdate/$1.md5 ]; then
-		str_md5_read=`cat $folderUpdate/$1.md5 | cut -d' ' -f1`
-	fi
-	if [ "$str_md5_calc" != "$str_md5_read" ]; then
-		echo "!!! FATAL ERROR: md5 does not match for file: $folderUpdate/$1" str_md5_calc = $str_md5_calc
-		# to inform u-boot upgrade tarball fail.
-		echo 1 > $bootcountfolder/tar_fail.txt
-		do_exit
-	fi
-}
-
-unzip_and_check_integrity() {
-	echo
-	echo "--- 3) Unzipping and checking integrity of upgrade package to $folderUpdate  ..."
-
-	# TODO: check return value of mkfs and mount??
-
-	# format cache partition
-	echo "--- formatting cache partition to " $cachepart ...
-	/sbin/mkfs.ext4 -F $cachepart
-	mount -t ext4 $cachepart $cachefolder
-
-	mkdir -p $folderUpdate
-	cd $folderUpdate
-
-	# debug only
-#	echo "--- copying to $folderUpdate ..."
-#	cp $sourcefolder/$entry update.tar.gz
-#	echo "--- unzipping $folderUpdate/update.tar.gz ..."
-#	tar xzf $folderUpdate/update.tar.gz --strip-components=1 2> /dev/null
-
-	echo "--- unzipping $sourcefolder/$entry ..."
-	# TODO: as files in the package are already tar,
-	# we should not use .tar.gz as in ch_fw_0.1_28.tar.gz.
-	# We better simply zip the tars together, but our initrd does not have zip tool.
-
-	# tar in initrd does not support --strip-components=1
-	tar xzf $sourcefolder/$entry
-	cd $folderUpdate/upgrade
-	mv * ../
-	cd .. && rmdir upgrade
-
-	echo "--- checking md5 integrity ..."
-	check_md5  $kernel_filename
-	check_md5  $rootfs_filename
-#	check_md5  $initrd_filename
-}
-
-
-###############################################################################
-# 4) Update rootfs into non-active bank
-###############################################################################
+    updateFileFound=0
+    echo
+    echo "--- Stage 2: Checking for upgrade file ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 2" >> /root/update.log
+
+    startString="ch_fw_"
+    lenString=`expr length "$startString"`
+
+    # Check for our .tar.gz file.
+    entry=`ls ${sourcefolder} | while read line; do echo ${line} | grep "^$startString.*.tar.gz$";done | head -n 1`
+
+    if [ "x$entry" == "x" ]; then
+        echo "--- No update file found. Writing error and exiting ..."
+        echo "=== $(date +%D_%T) Update Log: No update file found." >> /root/update.log
+        do_exit
+    fi
+
+    if [ -d "$sourcefolder/$entry" ]; then
+        echo "--- Found an incorrect entry (folder). Removing $entry ..."
+        rm -rf "$sourcefolder/$entry"
+        echo "--- Writing error log to $sourcepartition/update.log"
+        echo "=== $(date +%D_%T) Update Log: Update did not occur, incorrect file or folder ($entry) removed." >> /root/update.log
+        do_exit
+    fi
+
+    lenMatch=`expr match "$entry" "$startString"`
+    echo "--- Entry is ${entry}, startString = ${startString}, lenString = ${lenString}, lenMatch = ${lenMatch}"
+    echo "=== $(date +%D_%T) Update Log: Entry is ${entry}, startString = ${startString}, lenString = ${lenString}, lenMatch = ${lenMatch}" >> /root/update.log
+
+    if [ $lenMatch -eq $lenString ]; then
+        updateFileFound=1
+        break
+    fi
+
+    if [ $updateFileFound -eq 0 ]; then
+        echo "--- No update file was found. Exiting ..."
+        echo "=== $(date +%D_%T) Update Log: No update file was found." >> /root/update.log
+        echo 1 > $bootcountfolder/tar_fail.txt
+        do_exit
+    else
+        echo "--- Found update file! Using ${entry} ..."
+        echo "=== $(date +%D_%T) Update Log: Update file used is ${entry}" >> /root/update.log
+    fi  
+}
+
+########################################################################
+# Stage 3: Untar update file and verify md5sum.
+########################################################################
+
+updatefolder=$cachefolder/update
+kernelname="uImage"
+rootfsname="cherry.rootfs.tar"
+initrdname="cherry.initrd.tar"
+ifcversion="cherry-version"
+
+check_md5sum() {
+    # $1: Filename (kernelname, rootfsname, etc)
+    filesum=`md5sum $updatefolder/$1 | cut -d' ' -f1`
+    
+    # Get the stored md5sum from the upgrade/recovery directory.
+    if [ -f $updatefolder/$1.md5 ]; then
+        md5string=`cat $updatefolder/$1.md5 | cut -d' ' -f1`
+    else
+        echo "--- Fatal Error: MD5Sum file not found in tar."
+        echo "=== $(date +%D_%T) Update Log: Fatal Error: MD5Sum file not found." >> /root/update.log
+        do_exit
+    fi
+
+    # Check if the md5 of our file and the stored md5 do not match.
+    if [ "$md5string" != "$filesum" ]; then
+        echo "--- Fatal Error: MD5 hashes do not match. File Sum = ${filesum}, Stored Sum = ${md5string}"
+        echo "=== $(date +%D_%T) Update Log: Fatal Error: MD5 hashes do not match. File Sum = ${filesum}, Stored Sum = ${md5string}" >> /root/update.log
+        echo "=== $(date +%D_%T) Update Log: Deleting file ${entry}" >> /root/update.log        
+        rm -rf "$sourcefolder/$entry"
+        echo ""
+        echo "--- Exiting ..."
+        echo 1 > $bootcountfolder/tar_fail.txt
+        do_exit
+    fi
+}
+
+format_cache_partiion() {
+    echo
+    echo "--- Formatting Cache Partition ($cachepartition) ..."
+    /sbin/mkfs.ext4 -F $cachepartition
+    mount -t ext4 $cachepartition $cachefolder
+}
+
+untar_and_verify() {
+    echo 
+    echo "--- Stage 3: Extracting update and verifying integrity in $updatefolder ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 3" >> /root/update.log
+    format_cache_partiion
+
+    mkdir -p "$updatefolder"
+    cd "$updatefolder"
+
+    echo "--- Extracting $sourcefolder/$entry ..."
+    tar -xzf "$sourcefolder/$entry"
+    if [ $? -ne 0 ]; then
+        echo '--- Cannot extract this file. Is it really a .tar.gz?'
+        echo "=== $(date +%D_%T) Update Log: Fatal Error: Could not extract the file ${entry}" >> /root/update.log
+        echo "=== $(date +%D_%T) Update Log: Deleting file ${entry}" >> /root/update.log
+        rm -rf "$sourcefolder/$entry"
+        do_exit
+    fi
+
+    cd "$updatefolder/upgrade"
+    mv * ../
+    cd ../
+    rmdir upgrade
+
+    echo "--- Checking file integrity ..."
+    check_md5sum $kernelname
+    check_md5sum $rootfsname
+}
+
+########################################################################
+# Stage 4: Update rootfs
+########################################################################
+
+format_dest_partition() {
+    echo "--- Formatting destination partition $destpartition ..."
+    /sbin/mkfs.ext4 -F $destpartition
+    mount -t ext4 $destpartition $destfolder
+}
 
 do_update_rootfs() {
-	echo
-	echo "--- 4) Updating rootfs into non-active bank ..."
+    echo
+    echo "--- Stage 4: Updating rootfs ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 4" >> /root/update.log
+    format_dest_partition
+
+    echo "--- Update rootfs to destination partition ..."
+    cd "$destfolder"
+    tar -xf "$updatefolder/$rootfsname"
+    cd "$destfolder/rootfs"
+    mv * ../
+    cd .. && rmdir rootfs
+    cd
+}
+
+########################################################################
+# Stage 5: Update kernel
+########################################################################
+
+do_update_kernel() {
+    echo
+    echo "--- Stage 5: Updating the kernel ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 5" >> /root/update.log
+    dd if=$updatefolder/uImage of=$kernelpartition bs=256k
+}
+
+########################################################################
+# Stage 6: Post-Update (Copy User Data)
+########################################################################
 
-	# format the partition to be updated
-	echo "--- formatting destination partition" $destpart ...
-	/sbin/mkfs.ext4 -F $destpart
-	mount -t ext4 $destpart $destfolder
+ret_post_update=0
+ret_post_update_filename="/tmp/ret_post_update"
+post_update_file=$destfolder/root/do_post_update.sh
 
-	echo "--- update rootfs to" $destpart ...
-	cd $destfolder
-	# tar in initrd does not support --strip-components=1
-	tar xf $folderUpdate/cherry.rootfs.tar
-	cd $destfolder/rootfs
-	mv * ../
-	cd .. && rmdir rootfs
-	cd
-}
+do_copy_user_data() {
+    echo
+    echo "--- Stage 6: Post Update ..."
 
+    if [ -f $post_update_file ]; then
+        echo "--- Running do_post_update.sh"
+        $post_update_file
+    else
+        echo "--- There is no post update script."
+    fi
+
+    if [ -f $ret_post_update_filename ]; then
+        ret_post_update=`cat $ret_post_update_filename`
+        if [ $ret_post_update -ne 1 ]; then
+            echo "--- Fatal Error: Post update failure."
+            echo "=== $(date +%D_%T) Update Log:  Fatal Error: Post update failure." >> /root/update.log
+            do_exit
+        fi
+    fi
+}
+
+########################################################################
+# Stage 7: Remove intermediate update files.
+########################################################################
 
-###############################################################################
-# 5) Update kernel
-###############################################################################
+do_remove_files() {
+    echo
+    echo "--- Stage 7: Remove intermediate update files ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 7" >> /root/update.log
+    rm -f "$ret_post_update_filename"
+
+    case $purpose in
+        upgrade)
+            rm "$sourcefolder/$entry"
+            echo "--- Upgrade complete. Notifying u-boot ..."
+            echo 1 > "$bootcountfolder/tar_fail.txt"
+            ;;
+    esac
+}
+
+########################################################################
+# Magic
+########################################################################
 
-do_update_kernel_dtb() {
-	echo
-	echo "--- 5) Updating kernel ..."
-	dd if=$folderUpdate/uImage of=$kernel_part bs=256K
-}
+do_mount_upgrade
+check_updatefile
+untar_and_verify
+do_update_rootfs
+do_update_kernel
+do_copy_user_data
+do_remove_files
+do_unmount
 
+echo
+echo
+echo "--- Script completed."
+echo "=== $(date +%D_%T) Update Log: Updater has completed." >> /root/update.log
+exit 0#!/bin/sh
+
+############################################################################################
+#
+# BashBunny Firmware Recovery and Updater Script - v2.0 - Hak5 2017
+# 
+# Input Arguments:
+#   $1 : upgrade or recovery
+#   $2 : tarball source
+#   $3 : destination
+#
+# Notes:
+#   - /dev/nandf: udisk
+#   - /dev/nandd: rootfs
+#   - /dev/nandh: cache
+#   - u-boot not upgradable via Linux but via PhoenixSuit only.
+#
+#   Upgrade stages:
+#   1) Mount $2 (Source Partition) and /dev/nandh (Cache Partition) 
+#   2) Check if upgrade file exists (ch_fw_*.tar.gz) at udisk.
+#   3) Unpack upgrade file and verify md5sum.
+#   4) Update the rootfs
+#   5) Update the kernel
+#   6) Copy user data
+#   7) Remove intermerdiate files in /dev/nandh
+#   8) Remove the upgrade package from /mnt/update
+#   9) Unmount parititions and reboot.
+############################################################################################
 
-###############################################################################
-# 6) Post update: copy user data
-#    This is done by run /etc/init.d/do_post_update/sh in /mnt/rootfs_update
-###############################################################################
+purpose=$1
+sourcepartition=$2
+destpartition=$3
+bootcountpartition=/dev/nanda
+cachepartition=/dev/nandh
+kernelpartition=/dev/nandc
 
-ret_post_update=0
-ret_post_update_filename="/tmp/ret_post_update"
-post_update_file=$destfolder/root/do_post_update.sh
+sourcefolder=/root/udisk
+destfolder=/root/rootfs
+cachefolder=/root/cache
+bootcountfolder=/root/bootcount
+mkdir -p $sourcefolder
+mkdir -p $destfolder
+mkdir -p $cachefolder
+mkdir -p $bootcountfolder
 
-do_copy_user_data() {
-	echo
-	echo "--- 6) Post update: copy user data"
-	if [ -f $post_update_file ]; then
-		[[ "$purpose" == "recovery" ]] && mount -o sync /dev/nandf /root/udisk
-		echo "--- Post update: run $post_update_file"
-		$post_update_file
-	else
-		echo "--- post update file $post_update_file does not exist"
-	fi
-
-	# read file $ret_post_update_filename which is generated by post_update
-	if [ -f $ret_post_update_filename ]; then
-		ret_post_update=`cat $ret_post_update_filename`
-		if [ $ret_post_update -ne 1 ]; then
-			echo "--- !!! FATAL ERROR: post update failed."
-			do_exit
-		fi
-	fi
+touch /root/update.log
+echo " " >> /root/update.log
+
+echo "==============Starting Update==============" >> /root/update.log
+echo "--- We are doing ${purpose}, Source is ${sourcepartition}, Destination is ${destpartition}"
+echo
+
+do_clear_bootcounter() {
+    echo "--- Clearing boot counter ..."
+    /sbin/mkfs.fat -F 16 -n BOOTCNT $bootcountpartition
+    mount -t vfat $bootcountpartition $bootcountfolder
+    echo 0 > $bootcountfolder/bootcnt.txt
 }
 
+do_mount_upgrade() {
+    sync
+    umount $bootcountfolder
+    umount $cachefolder
+    umount $destfolder
+
+    echo "--- Stage 1: Mounting Partitions"
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 1" >> /root/update.log
+    case $purpose in
+        upgrade)
+            echo "--- We are upgrading firmware ..."
+            umount $sourcepartition
+            mount $sourcepartition $sourcefolder
+
+            # Clear the boot counter.
+            do_clear_bootcounter
+            ;;
+        recovery)
+            echo "--- We are recovering firmware ..."
+            sourcefolder=/root
+
+            # Clear the boot counter.
+            do_clear_bootcounter
+            ;;
+        *)
+            echo "--- Fatal Error: input is not upgrade or recovery! Exiting..."
+            exit 1
+            ;;
+    esac
+}
 
-###############################################################################
-# 7) If u-boot version is different, update u-boot
-###############################################################################
-
-check_and_update_uboot() {
-	echo
-	echo "--- 7) Cherry can not update u-boot"
+do_unmount() {
+    echo "--- Unmounting everything ..."
+    cd
+    sync
+    umount $bootcountpartition
+    umount $cachepartition
+    umount $destfolder
+
+    case $purpose in
+        upgrade)
+            umount $sourcepartition
+            ;;
+    esac
 }
 
+do_exit() {
+    echo "--- Exiting script and writing log file ..."
+    mv /root/update.log $sourcefolder
+    do_unmount
+    exit
+}
 
-###############################################################################
-# 8) Remove the intermediate files in non-active bank
-#    Remove upgrade package from /mnt/update (/dev/mmcblk0p4)
-###############################################################################
 
-do_remove_files() {
-	echo
-	echo "--- 8) Removing upgrade package files ..."
-	rm -f $ret_post_update_filename
-	# does not remove cache files
-#	rm -rf $folderUpdate
-
-	case $purpose in
-	upgrade)
-		# remove upgrade package from mass storage partition
-		rm $sourcefolder/$entry
-		echo "--- upgrade is done, notify u-boot tar is removed."
-		echo 1 > $bootcountfolder/tar_fail.txt
-		;;
-	recovery)
-		echo "--- does not remove recovery package files"
-		;;
-	esac
+########################################################################
+# Stage 2: Check if upgrade package exists in $sourcefolder.
+########################################################################
+
+check_updatefile() {
+    updateFileFound=0
+    echo
+    echo "--- Stage 2: Checking for upgrade file ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 2" >> /root/update.log
+
+    startString="ch_fw_"
+    lenString=`expr length "$startString"`
+
+    # Check for our .tar.gz file.
+    entry=`ls ${sourcefolder} | while read line; do echo ${line} | grep "^$startString.*.tar.gz$";done | head -n 1`
+
+    if [ "x$entry" == "x" ]; then
+        echo "--- No update file found. Writing error and exiting ..."
+        echo "=== $(date +%D_%T) Update Log: No update file found." >> /root/update.log
+        do_exit
+    fi
+
+    if [ -d "$sourcefolder/$entry" ]; then
+        echo "--- Found an incorrect entry (folder). Removing $entry ..."
+        rm -rf "$sourcefolder/$entry"
+        echo "--- Writing error log to $sourcepartition/update.log"
+        echo "=== $(date +%D_%T) Update Log: Update did not occur, incorrect file or folder ($entry) removed." >> /root/update.log
+        do_exit
+    fi
+
+    lenMatch=`expr match "$entry" "$startString"`
+    echo "--- Entry is ${entry}, startString = ${startString}, lenString = ${lenString}, lenMatch = ${lenMatch}"
+    echo "=== $(date +%D_%T) Update Log: Entry is ${entry}, startString = ${startString}, lenString = ${lenString}, lenMatch = ${lenMatch}" >> /root/update.log
+
+    if [ $lenMatch -eq $lenString ]; then
+        updateFileFound=1
+        break
+    fi
+
+    if [ $updateFileFound -eq 0 ]; then
+        echo "--- No update file was found. Exiting ..."
+        echo "=== $(date +%D_%T) Update Log: No update file was found." >> /root/update.log
+        echo 1 > $bootcountfolder/tar_fail.txt
+        do_exit
+    else
+        echo "--- Found update file! Using ${entry} ..."
+        echo "=== $(date +%D_%T) Update Log: Update file used is ${entry}" >> /root/update.log
+    fi  
+}
+
+########################################################################
+# Stage 3: Untar update file and verify md5sum.
+########################################################################
+
+updatefolder=$cachefolder/update
+kernelname="uImage"
+rootfsname="cherry.rootfs.tar"
+initrdname="cherry.initrd.tar"
+ifcversion="cherry-version"
+
+check_md5sum() {
+    # $1: Filename (kernelname, rootfsname, etc)
+    filesum=`md5sum $updatefolder/$1 | cut -d' ' -f1`
+    
+    # Get the stored md5sum from the upgrade/recovery directory.
+    if [ -f $updatefolder/$1.md5 ]; then
+        md5string=`cat $updatefolder/$1.md5 | cut -d' ' -f1`
+    else
+        echo "--- Fatal Error: MD5Sum file not found in tar."
+        echo "=== $(date +%D_%T) Update Log: Fatal Error: MD5Sum file not found." >> /root/update.log
+        do_exit
+    fi
+
+    # Check if the md5 of our file and the stored md5 do not match.
+    if [ "$md5string" != "$filesum" ]; then
+        echo "--- Fatal Error: MD5 hashes do not match. File Sum = ${filesum}, Stored Sum = ${md5string}"
+        echo "=== $(date +%D_%T) Update Log: Fatal Error: MD5 hashes do not match. File Sum = ${filesum}, Stored Sum = ${md5string}" >> /root/update.log
+        echo "=== $(date +%D_%T) Update Log: Deleting file ${entry}" >> /root/update.log        
+        rm -rf "$sourcefolder/$entry"
+        echo ""
+        echo "--- Exiting ..."
+        echo 1 > $bootcountfolder/tar_fail.txt
+        do_exit
+    fi
+}
+
+format_cache_partiion() {
+    echo
+    echo "--- Formatting Cache Partition ($cachepartition) ..."
+    /sbin/mkfs.ext4 -F $cachepartition
+    mount -t ext4 $cachepartition $cachefolder
+}
+
+untar_and_verify() {
+    echo 
+    echo "--- Stage 3: Extracting update and verifying integrity in $updatefolder ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 3" >> /root/update.log
+    format_cache_partiion
+
+    mkdir -p "$updatefolder"
+    cd "$updatefolder"
+
+    echo "--- Extracting $sourcefolder/$entry ..."
+    tar -xzf "$sourcefolder/$entry"
+    if [ $? -ne 0 ]; then
+        echo '--- Cannot extract this file. Is it really a .tar.gz?'
+        echo "=== $(date +%D_%T) Update Log: Fatal Error: Could not extract the file ${entry}" >> /root/update.log
+        echo "=== $(date +%D_%T) Update Log: Deleting file ${entry}" >> /root/update.log
+        rm -rf "$sourcefolder/$entry"
+        do_exit
+    fi
+
+    cd "$updatefolder/upgrade"
+    mv * ../
+    cd ../
+    rmdir upgrade
+
+    echo "--- Checking file integrity ..."
+    check_md5sum $kernelname
+    check_md5sum $rootfsname
+}
+
+########################################################################
+# Stage 4: Update rootfs
+########################################################################
+
+format_dest_partition() {
+    echo "--- Formatting destination partition $destpartition ..."
+    /sbin/mkfs.ext4 -F $destpartition
+    mount -t ext4 $destpartition $destfolder
 }
 
+do_update_rootfs() {
+    echo
+    echo "--- Stage 4: Updating rootfs ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 4" >> /root/update.log
+    format_dest_partition
+
+    echo "--- Update rootfs to destination partition ..."
+    cd "$destfolder"
+    tar -xf "$updatefolder/$rootfsname"
+    cd "$destfolder/rootfs"
+    mv * ../
+    cd .. && rmdir rootfs
+    cd
+}
+
+########################################################################
+# Stage 5: Update kernel
+########################################################################
+
+do_update_kernel() {
+    echo
+    echo "--- Stage 5: Updating the kernel ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 5" >> /root/update.log
+    dd if=$updatefolder/uImage of=$kernelpartition bs=256k
+}
+
+########################################################################
+# Stage 6: Post-Update (Copy User Data)
+########################################################################
+
+ret_post_update=0
+ret_post_update_filename="/tmp/ret_post_update"
+post_update_file=$destfolder/root/do_post_update.sh
+
+do_copy_user_data() {
+    echo
+    echo "--- Stage 6: Post Update ..."
+
+    if [ -f $post_update_file ]; then
+        echo "--- Running do_post_update.sh"
+        $post_update_file
+    else
+        echo "--- There is no post update script."
+    fi
+
+    if [ -f $ret_post_update_filename ]; then
+        ret_post_update=`cat $ret_post_update_filename`
+        if [ $ret_post_update -ne 1 ]; then
+            echo "--- Fatal Error: Post update failure."
+            echo "=== $(date +%D_%T) Update Log:  Fatal Error: Post update failure." >> /root/update.log
+            do_exit
+        fi
+    fi
+}
+
+########################################################################
+# Stage 7: Remove intermediate update files.
+########################################################################
+
+do_remove_files() {
+    echo
+    echo "--- Stage 7: Remove intermediate update files ..."
+    echo "=== $(date +%D_%T) Update Log: Starting Stage 7" >> /root/update.log
+    rm -f "$ret_post_update_filename"
+
+    case $purpose in
+        upgrade)
+            rm "$sourcefolder/$entry"
+            echo "--- Upgrade complete. Notifying u-boot ..."
+            echo 1 > "$bootcountfolder/tar_fail.txt"
+            ;;
+    esac
+}
+
+########################################################################
+# Magic
+########################################################################
 
 do_mount_upgrade
 check_updatefile
-unzip_and_check_integrity
+untar_and_verify
 do_update_rootfs
-do_update_kernel_dtb
+do_update_kernel
 do_copy_user_data
-check_and_update_uboot
 do_remove_files
 do_unmount
 
-
 echo
-echo "--- Done."
+echo
+echo "--- Script completed."
+echo "=== $(date +%D_%T) Update Log: Updater has completed." >> /root/update.log
 exit 0
